stdlib/Dates/src/io.jl

```diff
@@ -529,6 +529,27 @@ end
 
 # show
 
+for T in (
+    :Year, :Month, :Week, :Day,
+    :Hour, :Minute, :Second, :Millisecond, :Microsecond, :Nanosecond
+)
+    @eval function Base.show(io::IO, p::$T)
+        if get(io, :compact, false)
+            print(io, p)
+        else
+            print(io, $(string(T)), '(', p.value, ')')
```

Comment:
No need to call `string` here:
```julia
print(io, $T, '(', p.value, ')')
```

Reply:
I tried that, but the problem here is as follows (if I removed that string):

```
julia> import Dates

julia> show(Dates.Hour(1))
Dates.Hour(1)

julia> show(Hour(1))
ERROR: UndefVarError: Hour not defined
Stacktrace:
 [1] top-level scope at REPL[2]:1
```
```
julia> using Dates

julia> show(Dates.Hour(1))
Hour(1)

julia> show(Hour(1))
Hour(1)
```
but if I put the string there, then in both situations (import vs using) it will just show `Hour` instead of `Dates.Hour`.

Reply:
Try just `T` instead of `$T`

Reply:
The results you show above are correct, that's what should be happening

Reply:
Using the type is nice because it can show the module when it's required:
```julia
julia> import Dates

julia> function Base.show(io::IO, p::Dates.Hour)
           print(io, typeof(p), '(', p.value, ')')
       end

julia> show(Dates.Hour(2))
Dates.Hour(2)

julia> using Dates

julia> show(Dates.Hour(2))
Hour(2)
```

Reply:
hmm I suppose. I guess what I'm trying to compare it to is the new DateTime output code
```
function Base.show(io::IO, dt::DateTime)
    if get(io, :compact, false)
        print(io, dt)
    else
        y,m,d = yearmonthday(dt)
        h = hour(dt)
        mi = minute(dt)
        s = second(dt)
        ms = millisecond(dt)
        if ms == 0
            print(io, "DateTime($y, $m, $d, $h, $mi, $s)")
        else
            print(io, "DateTime($y, $m, $d, $h, $mi, $s, $ms)")
        end
    end
end
```
which will show "DateTime" regardless of how it was imported/etc

Reply:
I would say that "DateTime" is wrong. If the point of these show methods is to provide users a copy/paste friendly way of displaying the code then we should make sure what the paste will work.

Reply:
That's probably wrong too; there were changes to printing of types that removes the prefix when it's accessible from the current scope, and this probably predates that.

Reply:
Oh! You don't need to use `@eval` here at all:
```julia
function Base.show(io::IO, p::P) where P <: Union{Year, Month, Week, ...}
    if get(io, :compact, false)
        print(io, p)
    else
        print(io, P, '(', p.value, ')')
    end
end
```

Reply:
ah yes interesting I'll try that. I was just kind of copying some other code in the package that was referencing all those Period types using that for loop/eval style.

Reply:
yeah I realize that. I was just pointing out that with `using`, either way it would return "Hour" instead of "Dates.Hour"

Reply:
But what Curtis said is better yeah

Reply:
hm well perhaps that DateTime string should also be changed in a similar manner

Reply:
You may also want to consider making:
```julia
const PERIODS = (:Year, :Month, :Week, :Day, :Hour, :Minute, :Second, :Millisecond, :Microsecond, :Nanosecond)
```
which can also be used in "periods.jl"

Reply:
> I was just pointing out that with using, either way it would return "Hour" instead of "Dates.Hour"

Which is fine, because after `using Dates` you can access it as `Hour` instead of `Dates.Hour` :)